app/bundles/LeadBundle/Segment/Stat/ChartQuery/SegmentContactsLineChartQuery.php

```diff
@@ -0,0 +1,246 @@
+     */
+    public function getTotalStats(int $total)
+    {
+        $totalCountDateTo = $this->getTotalToDateRange($total);
+        // count array SUM and then reverse
+        // require start from end and  substract added/removed logs
+        $sums     = array_reverse(ArrayHelper::sub($this->getAddedEventLogStats(), $this->getRemovedEventLogStats()));
+        $totalSum = 0;
+        $totals   = array_map(function ($sum) use ($totalCountDateTo, &$totalSum) {
+            $totalSum += $sum;
+            $total = $totalCountDateTo - $totalSum;
+            if ($total > -1) {
+                return $total;
+            } else {
+                return 0;
+            }
+        }, $sums);
+
+        return array_reverse($totals);
+    }
+
+    /**
+     * Return total of contact to date end of graph.
+     *
+     * @param int $total
+     *
+     * @return int
+     */
+    private function getTotalToDateRange(int $total)
+    {
+        $queryForTotal = clone $this;
+        // try figure out total count in dateTo
+        $queryForTotal->setDateRange($this->dateTo, new \DateTime());
+
+        return $total - array_sum(ArrayHelper::sub($queryForTotal->getAddedEventLogStats(), $queryForTotal->getRemovedEventLogStats()));
+    }
+
+    /**
+     * Get data about add/remove from segment based on LeadEventLog.
+     *
+     * @param $action
+     *
+     * @return array
+     */
+    public function getDataFromLeadEventLog($action)
+    {
+        $actionInverted = ($action == 'added') ? 'removed' : 'added';
+
+        $filter              = [];
+        $filter['object']    = 'segment';
+        $filter['bundle']    = 'lead';
+        $filter['action']    = $action;
+        $filter['object_id'] = $this->segmentId;
+
+        $q = $this->prepareTimeDataQuery('lead_event_log', 'date_added', $filter);
+        $q->select('DATE_FORMAT(t.date_added, \''.$this->translateTimeUnit().'\') AS date, (COUNT(DISTINCT t.lead_id)) as count');
+
+        $subQuery = $this->connection->createQueryBuilder();
+        $subQuery->select('null')
+            ->from(MAUTIC_TABLE_PREFIX.'lead_event_log', 'el')
+            ->where(
+                $subQuery->expr()->andX(
+                    $subQuery->expr()->eq('el.lead_id', 't.lead_id'),
+                    $subQuery->expr()->eq('el.object', $subQuery->expr()->literal('segment')),
+                    $subQuery->expr()->eq('el.bundle', $subQuery->expr()->literal('lead')),
+                    $subQuery->expr()->eq('el.action', $subQuery->expr()->literal($actionInverted)),
+                    $subQuery->expr()->eq('el.object_id', $this->segmentId),
+                    $subQuery->expr()->eq('DATE_FORMAT(el.date_added, \''.$this->translateTimeUnit().'\')', 'DATE_FORMAT(t.date_added, \''.$this->translateTimeUnit().'\')'),
+                    $subQuery->expr()->gt('el.date_added', 't.date_added')
+                ));
+        $q->andWhere(sprintf('NOT EXISTS (%s)', $subQuery->getSQL()));
```

Comment:
@kuzmany we are hitting an issue with this subquery. It times out the segment detail page if someones adds a couple hundred thousand contacts and then removes them. If we understand this subquery correctly then it should hide/reduce the count of contacts that were added and removed within the same time frame. What we don't understand is why? I would want to see the spikes of added and removed contacts when debugging what happened with the segment. Also, if you'd change the timeframe then the data would show up anyway. Are you OK if we remove the subquery?

Reply:
Like you said. This subquery should cover to determine real number of subscribe/unsubscribed during the time frame. Then count just added and not remove after that, also removed count just removed and not-re-added after.
If we change it, we get different numbers. But It could change also total graph - we count it as total - (added - removed) https://github.com/mautic/mautic/blob/eee57aa627a649d8d74fe11882d47d3cea474262/app/bundles/LeadBundle/Segment/Stat/ChartQuery/SegmentContactsLineChartQuery.php#L80

But I can try test it on some graphs what happen without subquery.

Reply:
We actually think that the current implementation does not show the real data. We'd like to see the actually added and actually removed counts. Not the difference. Plus, removing that subquery would allow our users to load the segment detail page at all, so we see it as win-win.